cpp/src/replace/replace.cu

```diff
@@ -13,31 +16,528 @@
+                   iter replacement,
+                   Type* __restrict__ out_data)
+{
+  int tid = threadIdx.x;
+  int blkid = blockIdx.x;
+  int blksz = blockDim.x;
+  int gridsz = gridDim.x;
+
+  int start = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  for (int i=start; i<size; i+=step) {
+    out_data[i] = cudf::bit_is_set(in_valid, i)? in_data[i] : replacement[i];
+  }
+}
+
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+ *        `replace_nulls` with the appropriate data types.
+ */
+/* ----------------------------------------------------------------------------*/
+struct replace_nulls_column_kernel_forwarder {
+  template <typename col_type,
+            std::enable_if_t<cudf::is_fixed_width<col_type>()>* = nullptr>
+  void operator()(cudf::column_view const& input,
+                  cudf::column_view const& replacement,
+                  cudf::mutable_column_view& output,
+                  cudaStream_t stream = 0)
+  {
+    cudf::size_type nrows = input.size();
+    cudf::experimental::detail::grid_1d grid{nrows, BLOCK_SIZE};
+
+    replace_nulls<<<grid.num_blocks, BLOCK_SIZE, 0, stream>>>(nrows,
+                                                              input.data<col_type>(),
+                                                              input.null_mask(),
+                                                              replacement.data<col_type>(),
+                                                              output.data<col_type>());
+
+  }
+  template <typename col_type,
+            std::enable_if_t<not cudf::is_fixed_width<col_type>()>* = nullptr>
+  void operator()(cudf::column_view const& input,
+                  cudf::column_view const& replacement,
+                  cudf::mutable_column_view& output,
+                  cudaStream_t stream = 0) {
+    CUDF_FAIL("Non-fixed-width types are not supported for replace.");
+  }
+};
+
+
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+ *        `replace_nulls` with the appropriate data types.
+ */
+/* ----------------------------------------------------------------------------*/
+struct replace_nulls_scalar_kernel_forwarder {
+  template <typename col_type,
+            std::enable_if_t<cudf::is_fixed_width<col_type>()>* = nullptr>
+  void operator()(cudf::column_view const& input,
+                  cudf::scalar const& replacement,
+                  cudf::mutable_column_view& output,
+                  cudaStream_t stream = 0)
+  {
+    cudf::size_type nrows = input.size();
+    cudf::experimental::detail::grid_1d grid{nrows, BLOCK_SIZE};
+
+    using ScalarType = cudf::experimental::scalar_type_t<col_type>;
+    auto s1 = static_cast<ScalarType const&>(replacement);
```

Comment:
This should use a `scalar_device_view` instead of passing `s1.value()` (which incurs a device -> host memcpy).

Reply:
It is passing s1.value to thrust::make_constant_iterator because you requested that an iterator be used to consolidate the kernel functions. This seems to contradict that, so I am not really sure what you want me to do?

Reply:
You can make an iterator that dereferences from the `scalar_device_view`. I believe @trevorsm7 came up with a solution for this.

Reply:
Done.